File: challenge2_solution_power.py
import math


def check_1(num):
    """
    Return True is input <int> num is 2^N, False else
    Comparator *2 solution
    """
    cmp = 1
    while cmp <= num:
        if cmp == num:
            return True     # num is 2^N
        cmp *= 2
    
    return False    # num is not 2^N
    

def check_2(num):
    """
    Return True is input <int> num is 2^N, False else
    log2 solution
    """
    if num == 0: 
        return False
        
    log2 = math.log(num) / math.log(2)
    remainder = log2 - math.floor(log2)
    return not(remainder)
    

def check_3(num):
    """
    Return True is input <int> num is 2^N, False else
    int2bin solution, e.g.  8 = 0b1000
                                   ^^^ all zeros   
    """
    if num == 0: 
        return False

    return not "1" in bin(num)[3:]

    
if __name__ == "__main__":            
    for num in [128, 127, 63, 64, 1, 0]:
        s = "num={}; checks: ".format(num)
        for func in (check_1, check_2, check_3):
            s += "{}, ".format(func(num))
        print(s)
        

File: challenge2_solution_power.c
#include <stdio.h>
#include <math.h>


int check_1(int num) {
    /*
    Checks if num is 2^N
    Comparator *2 solution
    */
    int cmp = 1;
    while (cmp <= num) {
        if (cmp == num) return 1;     // num is 2^N
        cmp *= 2;
    }
    return 0;    // num is not 2^N    
}


int check_2(int num) {
    /*
    Checks if num is 2^N
    log2 solution
    */
    if (num == 0) return 0;

    float log2 = log(num) / log(2);
    if ((log2 - floor(log2)) == 0) return 1;  // num is 2^N

    return 0;                                 // num is not 2^N    
}


int check_3(int num) {
    /*
    Checks if num is 2^N
    Binary evaluation solution
    */
    if (num == 0) return 0;
    if (!(num & (num-1))) return 1;  // num is 2^N
    return 0;                        // num is not 2^N    
}



void main(void) {
    int nums[] = {128, 127, 63, 64, 1, 0};  // int array of test numbers
    int len = sizeof(nums) / sizeof(nums[0]);
    for(int i=0; i<len; i++) {
        printf("num=%d; checks: %d, %d, %d\n", nums[i], check_1(nums[i]), check_2(nums[i]), check_3(nums[i]));
    }
}